src/threshold.rs

```diff
@@ -0,0 +1,218 @@
+use cosmwasm_std::{
+    CheckedFromRatioError, ConversionOverflowError, Decimal, DivideByZeroError, Fraction,
+    OverflowError, StdError, Storage, Uint128,
+};
+use cw_storage_plus::Map;
+use std::convert::Infallible;
+use thiserror::Error;
+
+use crate::state::Stream;
+
+pub type Threshold = Uint128;
+
+#[derive(Error, Debug, PartialEq)]
+pub enum ThresholdError {
+    #[error(transparent)]
+    Std(#[from] StdError),
+
+    #[error("{0}")]
+    Overflow(#[from] OverflowError),
+
+    #[error("{0}")]
+    Infallible(#[from] Infallible),
+
+    #[error("{0}")]
+    DivideByZeroError(#[from] DivideByZeroError),
+
+    #[error("{0}")]
+    ConversionOverflowError(#[from] ConversionOverflowError),
+
+    #[error("{0}")]
+    CheckedFromRatioError(#[from] CheckedFromRatioError),
+
+    #[error("Threshold not reached")]
+    ThresholdNotReached {},
+
+    #[error("Threshold reached")]
+    ThresholdReached {},
+
+    #[error("Threshold not set")]
+    ThresholdNotSet {},
+
+    #[error("Min price can't be zero")]
+    ThresholdZero {},
+}
+pub const THRESHOLDS_STATE_KEY: &str = "thresholds";
+
+pub struct ThresholdState<'a>(Map<'a, u64, Threshold>);
+
+impl<'a> ThresholdState<'a> {
+    pub fn new() -> Self {
+        ThresholdState(Map::new(THRESHOLDS_STATE_KEY))
+    }
+    pub fn set_treshold_if_any(
+        &self,
+        stream: Stream,
```

Comment:
If possible borrow stream to reduce memory use
```suggestion
        stream: &Stream,
```